cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -448,6 +449,13 @@ Status CopyFiles(const std::vector<FileLocator>& sources,
                            destinations.size(), " paths.");
   }
 
+  RETURN_NOT_OK(::arrow::internal::OptionalParallelFor(
+      use_threads, static_cast<int>(sources.size()), [&](int i) {
+        auto dest_dir = internal::GetAbstractPathParent(destinations[i].path).first;
+        return dest_dir.empty() ? Status::OK()
+                                : destinations[i].filesystem->CreateDir(dest_dir);
```

Comment:
This may be very redundant, for example when copying lots of files into a single destination directory. If we had FileLocator::ToUri or similar we could maintain an unordered_set to ensure each directory is created only once

Reply:
Yes, this doesn't look terrific at all, especially on a remote filesystem. Why do you need to create target directories on the fly at all? That doesn't seem very useful, nor idiomatic.

Reply:
Well, I could just remove the directory creation. @nealrichardson ?

Reply:
`cp file.ext new/path` fails if `new/path` doesn't exist. `cp dir newdir` fails if `dir` is a directory. But `cp -r dir newdir` works and makes nested directories. 

So maybe the idiomatic solution is either (1) add a "recursive" argument, or (2) have a version that takes a single fs+FileSelector and a single destination fs+path, and that does the `cp -r` case.

The use case I'm interested in supporting is "copy the contents of this S3 bucket locally" and similar things.

Reply:
I'll add an argument to make directory creation optional, then

Reply:
I would simply remove directory creation. A separate function for recursive copy can be implemented if desired.